src/typings/formatters.ts

```diff
@@ -26,14 +28,100 @@ interface LogbackFormatterOptionsType {
     formatString: string,
 }
 
-type FormatterOptionsType = LogbackFormatterOptionsType;
+/**
+ * @property formatString A Yscope format string. The format string can include field-placeholders
+ * to insert and format any field of a JSON log event. A field-placeholder uses the following
+ * syntax:
+ * `{<field-name>[:<formatter-name>[:<formatter-options>]]}`
+ * - <field-name> (required) defines the key of the field whose value should replace the
+ * placeholder.
+ * - Nested fields can be specified using periods (`.`) to denote hierarchy. E.g., the field
+ * `{"a:" {"b": 0}}` may be denoted by `a.b`.
+ * - To denote a field-name with periods escape the periods with a backslashes.
+ * - <formatter-name> (optional) is the name of the formatter to apply to the value before
+ * inserting it into the string.
+ * - <formatter-options> (optional) defines any options for the formatter denoted by formatter-name.
+ *
+ * All three as may contain any character, except that colons (:), right braces (}),
+ * and backslashes (\) must be escaped with a backslash.
+ */
+interface YscopeFormatterOptionsType {
+    formatString: string,
+}
+
+type FormatterOptionsType = LogbackFormatterOptionsType | YscopeFormatterOptionsType;
 
 interface Formatter {
+
+    /**
+     * Formats the given log event.
+     *
+     * @param logEvent
+     * @return The formatted log event.
+     */
     formatLogEvent: (logEvent: LogEvent) => string
 }
 
+interface YScopeFieldFormatter {
+
+    /**
+     * Formats the given field.
+     *
+     * @param logEvent
+     * @return The formatted field.
+     */
+    formatField: (field: JsonValue) => string
+}
+
+/**
+ * Parsed field placeholder from a Yscope format string.
+ */
+type YScopeFieldPlaceholder = {
+    fieldNameKeys: string[],
+    fieldFormatter: Nullable<YScopeFieldFormatter>,
+}
+
+/**
+ * Type for list of currently supported Yscope field formatters.
+ */
+type YScopeFieldFormatterMap = {
+    [key: string]: new (options: Nullable<string>) => YScopeFieldFormatter;
+};
+
+// Patterns to assist parsing YScope format string. All patterns accept unescaped
+// matches only.
+/**
+ * Pattern to remove backlash.
+ */
+const BACKSLASH_REGEX = Object.freeze(/(?<!\\)\\/);
```

Comment:
@kirkrodrigues To represent two literal backslashes "\\", do we `\\\` or `\\\\`?
This Regex takes cares of `\\\` if i'm not wrong.
However, according to the `ANTLR Grammar` in our internal docs, it seems we need `\\\\`, which might not be possible with a single Regex pattern?
(Please correct me if I'm wrong with any of these. )

Reply:
We need `\\\\` (one escape per literal backslash).

Reply:
Sorry I might have misunderstood the regex in the first place. 😅 I'll revisit those as the first thing when I review this again.